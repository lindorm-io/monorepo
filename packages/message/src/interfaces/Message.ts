export interface IMessage {
  [key: string]: any;
}
