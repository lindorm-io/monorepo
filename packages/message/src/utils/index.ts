export * from "./default-create-message";
export * from "./default-generate-message";
export * from "./default-validate-message";
export * from "./get-topic-name";
export * from "./global";
