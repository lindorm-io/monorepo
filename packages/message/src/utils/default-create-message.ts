import { isFunction } from "@lindorm/is";
import { Constructor, DeepPartial } from "@lindorm/types";
import { IMessage } from "../interfaces";
import { globalMessageMetadata } from "./global";
import { parseField } from "./private";

export const defaultCreateMessage = <
  M extends IMessage,
  O extends DeepPartial<M> = DeepPartial<M>,
>(
  Message: Constructor<M>,
  options: O | M = {} as O,
): M => {
  const metadata = globalMessageMetadata.get(Message);
  const message = new Message();

  for (const field of metadata.fields) {
    (message as any)[field.key] = parseField(field, message, options);

    if ((message as any)[field.key]) continue;

    (message as any)[field.key] = isFunction(field.fallback)
      ? field.fallback()
      : field.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(message);
  }

  return message;
};
