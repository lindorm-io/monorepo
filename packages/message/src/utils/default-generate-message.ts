import { Constructor, DeepPartial } from "@lindorm/types";
import { randomBytes, randomInt, randomUUID } from "crypto";
import { IMessage } from "../interfaces";
import { MessageMetadata, MetaGenerated } from "../types";
import { globalMessageMetadata } from "./global";

const generator = (config: Omit<MetaGenerated, "target">): any => {
  switch (config.strategy) {
    case "date":
      return new Date();

    case "float":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return Math.random() * (config.max - config.min) + config.min;

    case "integer":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return randomInt(config.min, config.max);

    case "string":
      config.length = config.length ?? 32;
      return randomBytes(config.length).toString("base64url");

    case "uuid":
      return randomUUID();

    default:
      return null;
  }
};

const validator = (
  config: Omit<MetaGenerated, "target">,
  metadata: MessageMetadata,
): void => {
  const field = metadata.fields.find((item) => item.key === config.key);

  if (!field) {
    throw new Error(`Field not found for key: ${config.key}`);
  }

  if (!field.type) return;

  switch (config.strategy) {
    case "date":
      if (field.type === "date") break;
      throw new Error(`Invalid field type for date generation: ${field.type}`);

    case "float":
      if (field.type === "float") break;
      throw new Error(`Invalid field type for int generation: ${field.type}`);

    case "integer":
      if (field.type === "integer") break;
      throw new Error(`Invalid field type for int generation: ${field.type}`);

    case "string":
      if (field.type === "string") break;
      throw new Error(`Invalid field type for string generation: ${field.type}`);

    case "uuid":
      if (field.type === "uuid") break;
      throw new Error(`Invalid field type for uuid generation: ${field.type}`);

    default:
      break;
  }
};

export const defaultGenerateMessage = <M extends IMessage>(
  target: Constructor<M>,
  message: DeepPartial<M>,
): M => {
  const metadata = globalMessageMetadata.get(target);

  for (const config of metadata.generated) {
    if (message[config.key]) continue;

    validator(config, metadata);

    (message as any)[config.key] = generator(config);
  }

  return message as M;
};
