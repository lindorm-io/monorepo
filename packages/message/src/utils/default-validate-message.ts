import { Constructor } from "@lindorm/types";
import { z, ZodArray, ZodNumber, ZodRawShape, ZodString, ZodType } from "zod";
import { IMessage } from "../interfaces";
import { MetaField } from "../types";
import { globalMessageMetadata } from "./global";

const getValidator = (field: Omit<MetaField, "target">): ZodType<any> | undefined => {
  switch (field.type) {
    case "array":
      return z.array(z.any());

    case "bigint":
      return z.bigint();

    case "boolean":
      return z.boolean();

    case "date":
      return z.date();

    case "email":
      return z.string().email();

    case "enum":
      return z.nativeEnum(field.enum);

    case "float":
      return z.number();

    case "integer":
      return z.number().int();

    case "object":
      return z.object({}).passthrough();

    case "string":
      return z.string();

    case "url":
      return z.string().url();

    case "uuid":
      return z.string().uuid();

    default:
      return;
  }
};

export const defaultValidateMessage = <M extends IMessage>(
  target: Constructor<M>,
  message: M,
): void => {
  const metadata = globalMessageMetadata.get(target);
  const validators: ZodRawShape = {};

  for (const field of metadata.fields) {
    let validator = getValidator(field);

    if (!validator) continue;

    if (field.type && ["array", "number", "string"].includes(field.type) && field.min) {
      validator = (validator as ZodArray<any> | ZodNumber | ZodString).min(field.min);
    }

    if (field.type && ["array", "number", "string"].includes(field.type) && field.max) {
      validator = (validator as ZodArray<any> | ZodNumber | ZodString).max(field.max);
    }

    if (field.nullable) {
      validator = validator.nullable();
    }

    if (field.optional) {
      validator = validator.optional();
    }

    validators[field.key] = validator;
  }

  z.object(validators).passthrough().parse(message);

  if (metadata.schema) {
    metadata.schema.parse(message);
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnValidate");

  for (const hook of hooks) {
    hook.callback(message);
  }
};
