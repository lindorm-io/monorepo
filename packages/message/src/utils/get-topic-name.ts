import { kebabCase } from "@lindorm/case";
import { Constructor } from "@lindorm/types";
import { IMessage } from "../interfaces";
import { globalMessageMetadata } from "./global";

type Options = {
  namespace?: string | null;
  topic?: string | null;
};

export const getTopicName = <M extends IMessage>(
  MessageConstructor: Constructor<M>,
  message: M,
  options: Options,
): string => {
  const metadata = globalMessageMetadata.get(MessageConstructor);

  if (metadata.topic?.callback) {
    return metadata.topic.callback(message);
  }

  if (metadata.message.topic) return metadata.message.topic;
  if (options.topic) return options.topic;

  const namespace = metadata.message.namespace || options.namespace;
  const messageName = metadata.message.name || MessageConstructor.name;

  if (namespace === "system") {
    throw new Error("The 'system' namespace is reserved for internal use");
  }

  const n = namespace ? `${kebabCase(namespace)}.` : "";
  const e = kebabCase(messageName);

  const name = `${n}${e}`;

  if (name.length > 120) {
    throw new Error(`Topic name exceeds 120 characters: ${name}`);
  }

  return name;
};
