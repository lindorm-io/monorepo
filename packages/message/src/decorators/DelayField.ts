import { globalMessageMetadata } from "../utils";

export function DelayField(): PropertyDecorator {
  return function (target, key) {
    globalMessageMetadata.addField({
      target: target.constructor,
      key: key.toString(),
      decorator: "DelayField",
      enum: null,
      fallback: 0,
      max: null,
      min: 0,
      nullable: false,
      optional: false,
      schema: null,
      type: "integer",
    });
  };
}
