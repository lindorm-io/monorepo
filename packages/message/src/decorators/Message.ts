import { Constructor } from "@lindorm/types";
import { IMessage } from "../interfaces";
import { MessageDecoratorOptions } from "../types";
import { globalMessageMetadata } from "../utils";

export function Message(options: MessageDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    globalMessageMetadata.addMessage({
      decorator: "Message",
      name: options?.name || target.name,
      namespace: options?.namespace || null,
      target: target as unknown as Constructor<IMessage>,
      topic: options?.topic || null,
    });
  };
}
