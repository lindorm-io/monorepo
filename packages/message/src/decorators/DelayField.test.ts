import { globalMessageMetadata } from "../utils";
import { DelayField } from "./DelayField";
import { Message } from "./Message";

describe("DelayField Decorator", () => {
  test("should add metadata", () => {
    @Message()
    class DelayFieldDecoratorMessage {
      @DelayField()
      delay!: number;
    }

    expect(globalMessageMetadata.get(DelayFieldDecoratorMessage)).toMatchSnapshot();
  });
});
