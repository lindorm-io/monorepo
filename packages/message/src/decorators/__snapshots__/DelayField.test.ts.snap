// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelayField Decorator should add metadata 1`] = `
{
  "fields": [
    {
      "decorator": "DelayField",
      "enum": null,
      "fallback": 0,
      "key": "delay",
      "max": null,
      "min": 0,
      "nullable": false,
      "optional": false,
      "schema": null,
      "type": "integer",
    },
  ],
  "generated": [],
  "hooks": [],
  "message": {
    "decorator": "Message",
    "name": null,
    "namespace": null,
    "topic": null,
  },
  "priority": undefined,
  "schema": undefined,
  "topic": undefined,
}
`;
